File: modules/expo-foreground-service/android/src/main/java/expo/modules/foregroundservice/ExpoForegroundServiceModule.kt
package expo.modules.foregroundservice

import android.content.Context
import android.content.Intent
import android.os.Build
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoForegroundServiceModule : Module() {

    override fun definition() = ModuleDefinition {
        Name("ExpoForegroundService")

        Function("start") { sample: String ->
            val context = this@ExpoForegroundServiceModule.ctx
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                ForegroundService.startService(context, sample)
            } else {
                val intent = Intent(context, ForegroundService::class.java)
                intent.putExtra("sample", sample)
                context.startService(intent)
            }
        }

        Function("stop") {
            val context = this@ExpoForegroundServiceModule.ctx
            ForegroundService.stopService(context)
        }
    }

    private val ctx get(): Context = requireNotNull(appContext.reactContext)
}

File: modules/expo-foreground-service/src/ExpoForegroundServiceModule.ts
import { NativeModule, requireNativeModule } from "expo";

import type { Sample } from "@/typings/recording";

declare class ExpoForegroundServiceModule extends NativeModule {
  start: (sample: Sample["label"]) => void;
  stop: () => void;
}

export default requireNativeModule<ExpoForegroundServiceModule>("ExpoForegroundService");
